### PR TITLE
Vector3 Interop tests

### PR DESCRIPTION
**Pull request by sejongoh**
The native type of Vector3 is struct {float x,y,z} whose size is 12 bytes. RyuJit uses 16-byte register or stack location to store a Vector3 variable with the assumptions below. New testcases are added to check whether RyuJit properly handle interop cases.

RyuJIt back-end makes two assumptions about Vector3 types.

Assumption1: Vector3 type args passed in registers or on stack is rounded to POINTER_SIZE and hence on 64-bit targets it can be read/written as if it were TYP_SIMD16.

Assumption2: Vector3 args passed in registers (e.g. unix) or on stack have their upper 4-bytes being zero. Similarly Vector3 return type value returned from a method will have its upper 4-bytes zeroed out.

### CODE REVIEW

**Comment by sejongoh** [comment]
Under testing.

**Comment by sejongoh** [comment]
@RussKeldorph @CarolEidt @sivarv PTAL.

File: tests/src/JIT/SIMD/Vector3Interop_r.csproj [hunk 1]
```diff
@@ -0,0 +1,49 @@
...
+  <ItemGroup>
+    <Service Include="{82A7F48D-3B50-4B1E-B82E-3ADA8210C358}" />
+  </ItemGroup>
+  <ItemGroup>
+    <Compile Include="Vector3.cs" />
+  </ItemGroup>
+  <!--  <PropertyGroup>
+    <ProjectJson>$(JitPackagesConfigFileDirectory)threading+thread\project.json</ProjectJson>
+    <ProjectLockJson>$(JitPackagesConfigFileDirectory)threading+thread\project.lock.json</ProjectLockJson>
+  </PropertyGroup>-->
```
Comment (RussKeldorph) [inline tests/src/JIT/SIMD/Vector3Interop_r.csproj]: Why is this commented?  Also, I found that I had to include a ProjectReference to CMakeLists.txt to ensure the native portion is appropriately rebuilt.  See tests/src/JIT/Methodical/structs/systemvbringup/structinregs.csproj for instance.

File: tests/src/JIT/SIMD/CMakeLists.txt [hunk 2]
```diff
@@ -0,0 +1,10 @@
+cmake_minimum_required (VERSION 2.6)
+project (Vector3TestNative)
+include_directories(${INC_PLATFORM_DIR})
+set(SOURCES Vector3TestNative.cpp )
+
+# add the executable
+add_library (Vector3TestNative SHARED ${SOURCES})
+
+# add the install targets
+install (TARGETS Vector3TestNative DESTINATION bin)
```
Comment (RussKeldorph) [inline tests/src/JIT/SIMD/CMakeLists.txt]: Some of the other CMakeLists.txt files have a set_target_properties at the end with a comment claiming it's necessary for System V systems.  I'm not sure that's the case, but I just wanted to make sure you've confirmed this tests will work on non-Windows.

Comment (sejongoh) [inline tests/src/JIT/SIMD/CMakeLists.txt]: I copied this file from other Interop testcase and just changed the names. I got the native so library via build.sh and it worked well.

File: tests/src/JIT/SIMD/Vector3Interop_r.csproj (hunk 1 above)
Comment (sejongoh) [inline tests/src/JIT/SIMD/Vector3Interop_r.csproj]: Those lines are commented in tests/src/JIT/SIMD/*.csproj files. I am not sure if they are intentionally commented or can be removed.

File: tests/src/JIT/SIMD/Vector3TestNative.cpp [hunk 3]
```diff
@@ -0,0 +1,161 @@
...
+   Vector3 b;
+} DT;
+
+//
+// PInvoke native call for Vector3 size check
+//
+
+EXPORT(int) __stdcall nativeCall_PInvoke_CheckVector3Size() 
+{
+    printf("nativeCall_PInvoke_CheckVector3Size: sizeof(Vector3) == %d\n", (int)sizeof(Vector3));
```
Comment (RussKeldorph) [inline tests/src/JIT/SIMD/Vector3TestNative.cpp]: In my experience, the native console output is not synchronized with the managed output, and I needed to insert fflush(stdout) to have things better interleaved.  I could have been doing something wrong, but I thought I would mention it in case it's useful.

File: tests/src/JIT/SIMD/Vector3Interop.cs [hunk 4]
```diff
@@ -0,0 +1,218 @@
...
+    [DllImport(@"Vector3TestNative", CallingConvention = CallingConvention.StdCall)]
+    public static extern DT nativeCall_PInvoke_Vector3InStruct(DT d);
+
+    public static bool test() 
+    {
+
+        // Expected return value is 12 bytes.
+        if (nativeCall_PInvoke_CheckVector3Size() != 12) 
+        {
+            Console.Write("The size of native Vector3 type is not 12 bytes\n");
```
Comment (RussKeldorph) [inline tests/src/JIT/SIMD/Vector3Interop.cs]: Use Console.WriteLine instead of embedding newline characters in strings.

File: tests/src/JIT/SIMD/Vector3Interop_r.csproj (hunk 1 above)
Comment (RussKeldorph) [inline tests/src/JIT/SIMD/Vector3Interop_r.csproj]: Ok, I guess it's fine of other projects do it.  I'm less certain you can live without the CMakeLists.txt reference, though.  Please test incremental builds of this individual project when only the .cpp file is modified.

---
In reply to: [60309839](https://github.com/dotnet/coreclr/pull/4398#discussion_r60309839) [](ancestors = 60309839)

File: tests/src/JIT/SIMD/Vector3Interop.cs [hunk 5]
```diff
@@ -0,0 +1,218 @@
...
+    
+    public static bool test() {
+        int x = 1;
+        
+        nativeCall_RPInvoke_Vector3Arg(callBack_RPInvoke_Vector3Arg, x);
+        if (!result) 
+        {
+            Console.Write("RPInvoke Vector3Arg test failed\n");
+            return false;            
+        }
```
Comment (RussKeldorph) [inline tests/src/JIT/SIMD/Vector3Interop.cs]: What about an RP/Invoke that returns Vector3?  Is that interesting?

File: tests/src/JIT/SIMD/Vector3Interop_r.csproj (hunk 1 above)
Comment (sejongoh) [inline tests/src/JIT/SIMD/Vector3Interop_r.csproj]: I just have added the CMakeList.txt reference. In my case, build.sh built the cpp file incrementally well without the reference, but I am not sure about Windows.

File: tests/src/JIT/SIMD/Vector3Interop.cs (hunk 5 above)
Comment (sejongoh) [inline tests/src/JIT/SIMD/Vector3Interop.cs]: Yeah, it might have the same problem as the P/Invoke case. I will add it.

**Comment by RussKeldorph** [comment]
In general, they look like solid, positive test cases.  I was perhaps expecting a bit more nefarious code that tries to exploit potential weaknesses around boundary cases, but I'm not sure if there's much else that's interesting.  Perhaps more complex parameter combinations (e.g. (int, Vector3, string), etc.) or returning structs containing Vector3?  Your call if you think that's overboard.

Otherwise, LGTM modulo minor comments.

**Comment by sejongoh** [comment]
I tried to write testcases where two Vector3 objects are located contiguously and bad codegen could corrupt memory or read non-zero upper bytes. So far, since codegen for accessing Vector3 looks okay, I am not sure what complex cases could be effective. nativeCall_PInvoke_Vector3InStruct returns a structure containing two Vector3 objects. I will try to update some testcases to use complex parameter combinations.

**Comment by sejongoh** [comment]
@dotnet-bot test Ubuntu x64 Checked Build and Test

**Comment by sejongoh** [comment]
@sivarv Could you take a look at the testcases?

**Comment by sejongoh** [comment]
@dotnet-bot test this please

**Comment by sejongoh** [comment]
@dotnet-bot test this please

**Comment by sejongoh** [comment]
@dotnet-bot test Ubuntu x64 Checked Build and Test

**Comment by sivarv** [comment]
Looks good.

**Comment by sejongoh** [comment]
@RussKeldorph @sivarv Thank you for review!

File: tests/src/JIT/SIMD/Vector3Interop.cs [hunk 6]
```diff
@@ -0,0 +1,339 @@
...
+        Vector3 v1, 
+        [MarshalAs(UnmanagedType.LPStr)] string s, 
+        Vector3 v2)
+    {
+        Vector3 tmp = new Vector3(2, 2, 2);
+        
+        // sum = (1, 2, 3) dot (2, 2, 2) = 12
+        float sum0 = Vector3.Dot(v1, tmp);   
+        // sum = (10, 20, 30) dot (2, 2, 2) = 20 + 40 + 60 = 120
+        float sum1 = Vector3.Dot(v2, tmp);
```
Comment (sivarv) [inline tests/src/JIT/SIMD/Vector3Interop.cs]: One of the assumptions that we wanted to validate is that when native layer passes Vector3 args to managed, the upper 4 bytes are zero initialialized by native compiler.  In this test case we are doing dot product of a native layer passed Vector3 arg with a managed vector3 local.  We know managed local will have its upper 4 bytes zero initialized and hence dot product of v1.tmp or v2.tmp will give the right result.

To validate the above mentioned assumption please also compute the following dot products
v1.v1
v2.v2
v1.v2

File: tests/src/JIT/SIMD/Vector3Interop.cs [hunk 7]
```diff
@@ -0,0 +1,339 @@
...
+        
+    
+    static void callBack_RPInvoke_Vector3InStruct(DT v)
+    {   
+        Vector3 tmp = new Vector3(2, 2, 2);
+        
+        // sum0 = (2,3,4) dot (2,2,2) = 4 + 6 + 8 = 18
+        float sum0 = Vector3.Dot(v.a, tmp);
+        // sum1 = (11,21,31) dot (2,2,2) = 22 + 42 + 62 = 126
+        float sum1 = Vector3.Dot(v.b, tmp);
```
Comment (sivarv) [inline tests/src/JIT/SIMD/Vector3Interop.cs]: Here too please compute the following dot products

v.a, v.b
v.a, v.a
v.b, v.b

I would expect this case to be handled correctly by RyuJIt, since a Vector3 is read off memory, RyuJIt will generate 8 byte followed by 4 byte memory read to load a Vector3 into xmm reg.

File: tests/src/JIT/SIMD/Vector3Interop.cs [hunk 8]
```diff
@@ -0,0 +1,339 @@
...
+    [DllImport(@"Vector3TestNative", CallingConvention = CallingConvention.StdCall)]        
+    public static extern bool nativeCall_RPInvoke_Vector3InComplexStruct(
+        CallBackDelegate_RPInvoke_Vector3InComplexStruct callBack);
+    
+    static bool result = false;
+   
+    static void callBack_RPInvoke_Vector3Arg(
+        int i, 
+        Vector3 v1, 
+        [MarshalAs(UnmanagedType.LPStr)] string s,
```
Comment (sivarv) [inline tests/src/JIT/SIMD/Vector3Interop.cs]: Also please add some comments to each of the test cases to indicate which assumption(s) it is validating.

File: tests/src/JIT/SIMD/Vector3Interop.cs (hunk 6 above)
Comment (sivarv) [inline tests/src/JIT/SIMD/Vector3Interop.cs]: The above method signature on Linux would mean that both v1 and v2 will  be in registers (unix allows 6 reg params).  Please modify it to have some addition args before v2 so that v2 is passed on stack.

File: tests/src/JIT/SIMD/Vector3Interop.cs [hunk 9]
```diff
@@ -0,0 +1,339 @@
...
+                Console.Write("PInvoke Vector3Arg test failed\n");
+                return false;            
+            }
+        }
+        
+        // JIT crashes with this testcase.
+        // Disabled temporarily.
+        // {
+        //     // Expected return value = 1 + 2 + 3 = 6
+        //     Vector3 ret = nativeCall_PInvoke_Vector3Ret();
```
Comment (sivarv) [inline tests/src/JIT/SIMD/Vector3Interop.cs]: When we enable this test, please also update this test to compute DotProduct(ret, ret).

File: tests/src/JIT/SIMD/Vector3Interop.cs [hunk 10]
```diff
@@ -0,0 +1,339 @@
...
+    public static extern DT nativeCall_PInvoke_Vector3InStruct(DT d);
+
+    [DllImport(@"Vector3TestNative", CallingConvention = CallingConvention.StdCall)]
+    public static extern void nativeCall_PInvoke_Vector3InComplexStruct(ref ComplexDT cdt);
+    
+    public static bool test() 
+    {
+
+        // Expected return value is 12 bytes.
+        if (nativeCall_PInvoke_CheckVector3Size() != 12)
```
Comment (sivarv) [inline tests/src/JIT/SIMD/Vector3Interop.cs]: If native Vector3 size is 12 bytes,  a couple of questions here

1) Is native compiler rounding the size of native Vector3 struct to 16 bytes like x64 CLR does?
2) While passing a native vector3 struct as arg, just 12 bytes passed or 16 bytes?
If it is just 12 bytes, in RPInvoke scenario is RyuJIT reading past the end of struct?  If it is 16 bytes, is native compiler ensuring that the upper 4 bytes of the struct are zero initialized?  Specifically does native compiler spec (on x64 windows and unix) mention that it would zero either round the size to 16 bytes while passing as arg or zero init upper 4 bytes to zero?

**Comment by tijoytom** [comment]
@sejongoh @sivarv @RussKeldorph @yizhang82  
I am trying to see if it makes sense to move this test to Interop fodler, we have plans to move interop tests to it's own repository and use it as the common set of test across platforms. Is these tests very specific to RyuJIT ?

**Comment by sejongoh** [comment]
@tijoytom I think that the testscase can be moved to Interop. @sivarv and @CarolEidt have a better understanding than I.

**Comment by sivarv** [comment]
@tijoytom - these tests are specific to RyuJIT as they are validating the assumptions that RyuJIT made w.r.t to SIMD Vector3 type.  I would suggest to leave them under JIT folder.

**Comment by CarolEidt** [comment]
I agree that these are best kept as JIT-specific.

File: tests/src/JIT/SIMD/Vector3Interop.cs (hunk 10 above)
Comment (sejongoh) [inline tests/src/JIT/SIMD/Vector3Interop.cs]: The native passes 12 bytes.

```
  ; Vector3 arg
  000008F5: 83 EC 0C           sub         esp,0Ch 
  000008F8: 8B C4              mov         eax,esp
  000008FA: 8B 4D F0           mov         ecx,dword ptr [ebp-10h]
  000008FD: 89 08              mov         dword ptr [eax],ecx
  000008FF: 8B 55 F4           mov         edx,dword ptr [ebp-0Ch]
  00000902: 89 50 04           mov         dword ptr [eax+4],edx
  00000905: 8B 4D F8           mov         ecx,dword ptr [ebp-8]
  00000908: 89 48 08           mov         dword ptr [eax+8],ecx
  ; str arg
  0000090B: 8B 15 00 00 00 00  mov         edx,dword ptr [?str@?1??nativeCall_RPInvoke_Vector3Arg@@9@4@4PADA] 
  00000911: 52                 push        edx
  ; Vector3 arg
  00000912: 83 EC 0C           sub         esp,0Ch
  00000915: 8B C4              mov         eax,esp
  00000917: 8B 4D E4           mov         ecx,dword ptr [ebp-1Ch]
  0000091A: 89 08              mov         dword ptr [eax],ecx
  0000091C: 8B 55 E8           mov         edx,dword ptr [ebp-18h]
  0000091F: 89 50 04           mov         dword ptr [eax+4],edx
  00000922: 8B 4D EC           mov         ecx,dword ptr [ebp-14h]
  00000925: 89 48 08           mov         dword ptr [eax+8],ecx
  ; int arg
  00000928: 8B 55 E0           mov         edx,dword ptr [ebp-20h]
  0000092B: 52                 push        edx
  ; call
  0000092C: FF 55 08           call        dword ptr [ebp+8]
```

RyuJit reads only 12 bytes and the upper 4bytes are initialized to zero when loading the value to register.

```
; The address of v is stored in V00. The value of V0 is loaded to a 128-bit register and reg[127:96] is initialized to zero. So, the dot product result is correct.
Generating: N019 (  1,  1) [000015] ------------             *  lclVar    byref  V00 arg0         u:2 rcx (last use) REG rcx RV $80
Generating: N021 (  7,  5) [000018] ---XG-------             *  indir     simd12 REG mm1
							V00 in reg rcx is becoming dead  [000015]
							Live regs: 00000002 {rcx mm0} => 00000000 {mm0}
							Live vars: {V00 V01} => {V01}
							Byref regs: 00000002 {rcx} => 00000000 {}
IN000b:        vmovss   ymm2, dword ptr [rcx+8]   ; --> upper 64 bits are initialized to zero
IN000c:        vmovsd   ymm1, qword ptr [rcx]     ; --> upper 96 bits are initialized to zero
IN000d:        vshufps  ymm1, ymm2, 68            ; 68 = 44h ymm1 <= ymm2[63:32]::ymm2[31:00]::ymm1[63:32] ::ymm1[31:0]
Generating: N023 (  3,  2) [000016] ------------             *  lclVar    simd16 V01 loc0         u:3 mm0 (last use) REG mm0 RV $201
Generating: N025 ( 11,  8) [000019] ---XG-------             *  simd      float  float Dot REG mm6 $102
							V01 in reg mm0 is becoming dead  [000016]
							Live regs: 00000000 {mm0} => 00000000 {}
							Live vars: {V01} => {}
IN000e:        vmovaps  ymm6, ymm1
IN000f:        vdpps    ymm6, ymm0, 241
Generating: N027 ( 15, 13) [000021] DA-XG-------             *  st.lclVar float  V02 loc1         d:3 mm6 REG mm6 RV
```

I will try to find spec.

Comment (sivarv) [inline tests/src/JIT/SIMD/Vector3Interop.cs]: The above is on Windows and a reference to copy of Vector3 is passed by native caller.  Managed callee would have to indirect on addr received as arg and RyuJIT would read just 12 bytes.

What about on x64 Unix?  Vector3 could be passed either in two registers or on stack.  If passed in registers, RyuJIT prolog homes Vector3 arg to a stack location. As per my understanding RyuJIT is homing 16 bytes and assumes upper 4 bytes are zero initialized.  Similarly, what happens when Vector3 arg is passed on stack, does RyuJIT generated code try to read 12 or 16 bytes?

**Comment by sejongoh** [comment]
On Linux, the following RPInvoke case fails. The second Vector3 object, v3f32_mem0 is assigned onto stack, and its z value is wrong. The testcase passes on Windows.

v3f32_mem0 is assigned (10,20,30) on the native side. I checked the values on stack right before call instruction and they look okay.

```
; v3f32_mem0 argument values on stack right before call
0x7fffffffd108: 0x41200000  (10.0)    0x41a00000  (20.0)     0x41f00000 (30.0)      0x40c00000 (local)
```

On the managed side, the code looks okay but v2f32_mem0.z has a wrong value, '1'.
```
movss    xmm0, dword ptr [rbp+18H]  ; xmm0 <- v2f32_mem0.x; rbp+18H == AddrOf(v2f32_mem0)
...
lea      rdi, bword ptr [rbp+18H]
movss    xmm0, dword ptr [rdi+4]    ; xmm0 <- v2f32_mem0.y; rbp+18H == AddrOf(v2f32_mem0)
...
lea      rdi, bword ptr [rbp+18H]
movss    xmm0, dword ptr [rdi+8]    ; xmm0 <- v2f32_mem0.z; rbp+18H == AddrOf(v2f32_mem0)
```

Native side
```

//
// RPInvoke native call for Vector3 argument
//
typedef void (__stdcall *CallBack_RPInvoke_Vector3Arg_Unix)(
    Vector3 v3f32_xmm0, 
    float   f32_xmm2,
    float   f32_xmm3,
    float   f32_xmm4,
    float   f32_xmm5,
    float   f32_xmm6,
    float   f32_xmm7,
    float   f32_mem0,
    Vector3 v3f32_mem0);

EXPORT(void) __stdcall nativeCall_RPInvoke_Vector3Arg_Unix(
  CallBack_RPInvoke_Vector3Arg_Unix notify)
{
    Vector3 v1, v2;
    v1.x = 1; v1.y = 2; v1.z = 3;
    v2.x = 10; v2.y = 20; v2.z = 30;
    float f0 = 0, f1 = 1, f2 = 2, f3 = 3, f4 = 4, f5 = 5, f6 = 6;
    notify(
        v1,  
        f0,
        f1,
        f2,
        f3,
        f4,
        f5,
        f6,
        v2);
} 
```

Callback at managed side
```
    static void callBack_RPInvoke_Vector3Arg_Unix(
        Vector3 v3f32_xmm0, 
        float   f32_xmm2,
        float   f32_xmm3,
        float   f32_xmm4,
        float   f32_xmm5,
        float   f32_xmm6,
        float   f32_xmm7,
        float   f32_mem0,
        Vector3 v3f32_mem0)
    {
        Vector3 tmp = new Vector3(2, 2, 2);
        
        // sum = (1, 2, 3) dot (1, 2, 3) = 14
        float sum0 = Vector3.Dot(v3f32_xmm0, v3f32_xmm0);   
        // sum = (10, 20, 30) dot (10, 20, 30) = 1400
        float sum1 = Vector3.Dot(v3f32_mem0, v3f32_mem0);       
        // sum = (1, 2, 3) dot (10, 20, 30) =140
        float sum2 = Vector3.Dot(v3f32_xmm0, v3f32_mem0);       
        
        Console.WriteLine("callBack_RPInvoke_Vector3Arg_NonWindows:");
        Console.WriteLine("    {0}, {1}, {2}", v3f32_xmm0.X, v3f32_xmm0.Y, v3f32_xmm0.Z);
        Console.WriteLine("    {0}, {1}, {2}", v3f32_mem0.X, v3f32_mem0.Y, v3f32_mem0.Z);
        Console.WriteLine("    Sum0, 1,2 = {0}, {1}, {2}", sum0, sum1, sum2);
        
        result = (sum0 == 14) && (sum1 == 1400) && (sum2 == 140);
    }    
```

**Comment by sejongoh** [comment]
I tried COMPlus_JitHalt=nativeCall_RPInvoke_Vector3Arg_Unix to debug the generated code but it didn't work. Did I miss anything?

**Comment by CarolEidt** [comment]
I have not been successful in using COMPlus_JitHalt on non-windows machines (though perhaps I'm doing something wrong as well). What I do is to figure out which method it is in the compile list, and set a conditional breakpoint for Compiler::jitTotalMethodCompiled == N at whatever point in the compile I'm interested. If you want to debug the generated code, set the breakpoint for the return from CodeGen::genGenerateCode and then set a breakpoint for the value pointed to by methodCodePointer.
